test/e2e/hsts_policy_test.go

```diff
@@ -0,0 +1,183 @@
+	}
+	defer func() {
+		// this will cleanup all components in this namespace
+		if err := kclient.Delete(context.TODO(), ns); err != nil {
+			t.Fatalf("failed to delete test namespace %s: %v", ns.Name, err)
+		}
+	}()
+
+	// Create pod
+	echoPod := buildEchoPod("hsts-policy-echo", ns.Name)
+	if err := kclient.Create(context.TODO(), echoPod); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", ns, echoPod.Name, err)
+	}
+
+	// Create service
+	echoService := buildEchoService(echoPod.Name, ns.Name, echoPod.ObjectMeta.Labels)
+	if err := kclient.Create(context.TODO(), echoService); err != nil {
+		t.Fatalf("failed to create service %s/%s: %v", echoService.Namespace, echoService.Name, err)
+	}
+
+	// Create a route that should fail the HSTS policy validation
+	t.Logf("creating an invalid route at %s", time.Now().Format(time.StampMilli))
+	invalidRoute := buildRouteWithHSTS("invalid-route", echoPod.Namespace, echoService.Name, "invalid-route."+appsDomain, "max-age=99999999")
+	if err := kclient.Create(context.TODO(), invalidRoute); err == nil {
+		t.Fatalf("failed to reject an invalid route %s/%s, max-age 99999999", invalidRoute.Namespace, invalidRoute.Name)
+	} else {
+		t.Logf("rejected an invalid route at %s: %s/%s with annotation %s: %v", time.Now().Format(time.StampMilli), invalidRoute.Namespace, invalidRoute.Name, invalidRoute.Annotations, err)
+	}
+
+	// Create a route that should pass the HSTS policy validation
+	exampleHeader := "max-age=0;preload;includesubdomains"
+	t.Logf("creating a valid route at %s", time.Now().Format(time.StampMilli))
+	validRoute := buildRouteWithHSTS("valid-route", echoPod.Namespace, echoService.Name, "valid-route."+appsDomain, exampleHeader)
+	if err := kclient.Create(context.TODO(), validRoute); err != nil {
+		t.Fatalf("failed to create a valid route %s/%s: %v", validRoute.Namespace, validRoute.Name, err)
+	} else {
+		t.Logf("created a valid route at %s: %s/%s with annotation %s", time.Now().Format(time.StampMilli), validRoute.Namespace, validRoute.Name, validRoute.Annotations)
+	}
+
+	// Wait for route service to respond, and when it does, check for the header that should be there
+	if err := wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
+		resultHeader, code, err := getHttpHeader(validRoute)
+		if code != http.StatusOK && err != nil {
+			return false, nil
+		}
+		header := resultHeader.Get("strict-transport-security")
+		if header == "" {
+			t.Logf("request to %s got no HSTS header: %s (%d) %v, trying again....", validRoute.Spec.Host, resultHeader, code, err)
+			return false, nil
+		}
+		if strings.Compare(header, exampleHeader) != 0 {
+			return false, fmt.Errorf("request to %s got wrong HSTS header, should be %s, not %s", validRoute.Spec.Host, exampleHeader, header)
+		}
+		t.Logf("request to %s got correct HSTS header: %s", validRoute.Spec.Host, header)
+		return true, nil
+	}); err != nil {
+		t.Fatalf("failed to find header %s: %v", exampleHeader, err)
+	}
```

Comment:
```suggest
	httpClient := &http.Client{
		Timeout: 10 * time.Second,
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
	}

	// Wait for route service to respond, and when it does, check for the header that should be there
	if err := wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
		resp, err := httpClient.Get("https://" + validRoute.Spec.Host)
		if err != nil {
			t.Logf("GET %s failed: %v, retrying...", validRoute.Spec.Host, err)
			return false, nil
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			t.Logf("GET %s failed: status %v, expected %v, retrying...", validRoute.Spec.Host, resp.StatusCode, http.StatusOK)
			return false, nil // retry on 503 as pod/service may not be ready
		}
		if header := resp.Header.Get("strict-transport-security"); header != exampleHeader {
			return false, fmt.Errorf("expected %q, got %q", exampleHeader, header)
		}
		return true, nil
	}); err != nil {
		t.Fatalf("failed to find header %s: %v", exampleHeader, err)
	}
```

Reply:
I made some changes, would like to keep the function, but otherwise incorporated your suggestions.

Reply:
OK, my issue was with the function (there are no other callers, not very generic, would be better done at the call site), but let's leave it for now.